narwhal/executor/src/subscriber.rs

```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
                     cert.digest()
                 );
-                let batch = self.fetch_payload(*digest, *worker_id, workers).await;
-                batches.push(batch);
+                output_batches.push(batch.clone());
             }
-            subscriber_output.batches.push((output_cert, batches));
+            subscriber_output
+                .batches
+                .push((output_cert, output_batches));
         }
-
         subscriber_output
     }
 
-    /// Fetches single payload from network
-    /// This future performs infinite retries and blocks until Batch is available
+    /// Bulk fetches payload from workers
+    /// This future performs infinite retries and blocks until all batches are available
     /// As an optimization it tries to download from local worker first, but then fans out
     /// requests to remote worker if not found locally
-    #[instrument(level = "debug", skip_all, fields(digest = % digest, worker_id = % worker_id))]
-    async fn fetch_payload(
+    async fn fetch_batches_from_worker(
         &self,
-        digest: BatchDigest,
-        worker_id: WorkerId,
+        batch_digests: HashMap<WorkerId, HashSet<BatchDigest>>,
         workers: Vec<NetworkPublicKey>,
-    ) -> Batch {
-        if let Some(payload) = self.try_fetch_locally(digest, worker_id).await {
-            let batch_fetch_duration = payload.metadata.created_at.elapsed().as_secs_f64();
-            self.metrics
-                .batch_execution_latency
-                .observe(batch_fetch_duration);
+    ) -> HashMap<BatchDigest, Batch> {
+        let mut fetched_batches = HashMap::new();
+
+        for (worker_id, digests) in batch_digests {
             debug!(
-                "Batch {:?} took {} seconds to be fetched for execution since creation",
-                payload.digest(),
-                batch_fetch_duration
+                "Attempting to fetch {} digests from {} worker_{worker_id}'s",
+                digests.len(),
+                workers.len()
             );
-            return payload;
-        }
-        let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
-        let mut stagger = Duration::from_secs(0);
-        let mut futures = vec![];
-        for worker in workers {
-            let future = self.fetch_from_worker(stagger, worker, digest);
-            futures.push(future.boxed());
-            // TODO: Make this a parameter, and also record workers / authorities that are down
-            //       to request from them batches later.
-            stagger += Duration::from_millis(200);
+            // TODO: Can further parallelize this by worker_id if necessary.
+            // Only have one worker for now so will leave this for a future
+            // optimization.
+            let mut remaining_digests = digests.clone();
+
+            let local_batches = self
+                .try_fetch_locally(remaining_digests.clone().into_iter().collect(), worker_id)
```

Comment:
Maybe limit the size of the vec of remaining_digests here instead of limiting it in the handler?

Reply:
There is no bad requester to worry about since this is between primary and worker who trust each other.